R/cut.R

```diff
@@ -0,0 +1,129 @@
+#' so on.
+#' @param x A crunch variable of the class NumericVariable
+#' @param breaks Either a numeric vector of two or more unique cut points
+#' or a single number (greater than or equal to 2) giving the number of intervals
+#' into which x is to be cut.
+#' @param name The name of the resulting case variable as a character string.
+#' @param labels labels for the levels of the resulting category.
+#' By default, labels are constructed using interval notation.
+#' If labels = FALSE, simple integer codes are returned instead of a factor.
+#' @param include.lowest logical, indicating if an `x[i]` equal to the lowest
+#' (or highest, for right = FALSE) `breaks` value should be included.
+#' @param right logical, indicating if the intervals should be closed on the right
+#' (and open on the left) or vice versa.
+#' @param dig.lab	integer which is used when labels are not given.
+#' It determines the number of digits used in formatting the break numbers.
+#' @param ordered_result	Ignored.
+#' @param ... further arguments passed to [makeCaseVariable]
+#' @return a Crunch [`VariableDefinition`]. Assign it into the dataset to create
+#' it as a derived variable on the server.
+#' @export
+#' @examples
+#' \dontrun{
+#' ds <- loadDataset("mtcars")
+#' ds$cat_var <- cut(ds$mpg, 3, variableName = "new_var")
+#' }
+#'
+setMethod("cut", "NumericVariable", function(x,
+                                             breaks,
+                                             name,
+                                             labels = NULL,
+                                             include.lowest = FALSE,
+                                             right = TRUE,
+                                             dig.lab = 3,
+                                             ordered_result = FALSE, ...){
+    if (missing(name)) {
+        halt("Must provide the name for the new variable")
+    }
+    if (length(breaks) == 1L) {
+        if (is.na(breaks) || breaks < 2L) {
+            halt("invalid number of intervals")
+        }
+        nb <- as.integer(breaks + 1) # one more than #{intervals}
+        rx <- c(min(x, na.rm = TRUE), max(x, na.rm = TRUE))
+        dx <- diff(rx)
+        if (dx == 0) {
+            dx <- abs(rx[1L])
+            breaks <- seq.int(rx[1L] - dx/1000, rx[2L] + dx/1000,
+                              length.out = nb)
+        } else {
+            breaks <- seq.int(rx[1L], rx[2L], length.out = nb)
+            breaks[c(1L, nb)] <- c(rx[1L] - dx/1000, rx[2L] + dx/1000)
+        }
+    } else {
+        breaks <- sort.int(as.double(breaks))
+        nb <- length(breaks)
+    }
+    if (anyDuplicated(breaks)) {
+        halt(sQuote("breaks"), " are not unique")
+    }
+    if (is.null(labels)) { # Autogenerate labels if not supplied
+        labels <- generateCutLabels(dig.lab, breaks, nb, right, include.lowest)
+    } else if (length(labels) != nb - 1L) {
+        halt("lengths of ", sQuote("breaks"), " and ", sQuote("labels"), " differ")
```

Comment:
This one puzzled me. The message says that the problem is that the lengths of breaks and labels differs, but in order for it to be valid, apparently you have to have one more "break" than "label"! Let's give a better message and clearer instructions in the man page.

Some other considerations: 

* Should we allow "labels" to be full category definitions? There are other attributes that our data type supports.
* Should we add an argument or something that changes the behavior for extreme values, i.e. makes categories for them? So that the correct number of "labels" to provide would be length(breaks) + 1, not -1. Looking at the base implementation, 

```
> cut(1:20, c(4, 8, 13))
 [1] <NA>   <NA>   <NA>   <NA>   (4,8]  (4,8]  (4,8]  (4,8]  (8,13] (8,13]
[11] (8,13] (8,13] (8,13] <NA>   <NA>   <NA>   <NA>   <NA>   <NA>   <NA>  
Levels: (4,8] (8,13]
```

my low and high extremes are just "NA", but we could treat them as different categories. Even without adding an argument about that, we could make them be separate categories (missing: true) that the user could still work with after creating the variable.

Reply:
I agree the message is a bit confusing, so I changed it to talk about "resulting categories " which is what should match the number of labels. It's just a bit confusing because `breaks` refers to the start and end points of the category, which is always one more than the number of categories. 

I like the idea of full category definitions but my intuition about this function is that we should try to keep it fairly simple and close to the base functions. Maybe we can hold off including the category definitions and see users request that functionality?

For the extreme value behaviour, I think there's kind of an argument either way. It seems like there's a use case for either distinguishing between high-value NA, and low-value NA, but on the other hand the user could think about numbers which fall outside of the range just as "unspecified" in which case they would want the same value.  I'm not sure an extra argument makes a lot of sense because it's more work than just specifying the breaks to capture those extreme values. For instance: 
```
cut(1:1000, 
       breaks = (5, 10, 100), 
       labels = c("tiny", "small", "medium", "large"),
       code_extreme = TRUE)
``` 
would be equivalent to 
```
cut(1:1000,
      breaks = (-10000, 5, 10, 100, 10000),
      labels = c("tiny", "small", "medium", "large")
```
but the former is more verbose and requires the user to be familiar with another argument.

Reply:
Fine to defer full category definitions and extreme behavior and see what users think.